File: mobile/customer-app/app/src/main/java/com/sukashawarma/customer/ui/components/States.kt
package com.sukashawarma.customer.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.sukashawarma.customer.data.api.GatewayError

@Composable
fun MemuatState(modifier: Modifier = Modifier) {
    Column(
        modifier = modifier.fillMaxWidth().padding(32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        CircularProgressIndicator(color = MaterialTheme.colorScheme.primary)
    }
}

@Composable
fun EmptyState(
    judul: String,
    penjelasan: String,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier.fillMaxWidth().padding(32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(judul, style = MaterialTheme.typography.titleLarge)
        Text(
            penjelasan,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center
        )
    }
}

/**
 * Menerjemahkan galat gateway jadi kalimat yang bisa ditindak pelanggan.
 *
 * Fungsi murni supaya bisa diuji tanpa menyalakan Compose, dan supaya tidak
 * ada layar yang diam-diam mengarang kalimatnya sendiri.
 */
fun pesanGalat(error: GatewayError): String = when (error) {
    is GatewayError.Jaringan ->
        "Tidak bisa terhubung. Periksa koneksi internetmu, lalu coba lagi."
    is GatewayError.Server ->
        "Layanan sedang bermasalah. Coba lagi sebentar lagi."
    GatewayError.SesiTidakSah ->
        "Sesimu sudah berakhir. Masuk lagi untuk melanjutkan."
    is GatewayError.Kode -> error.pesan
}

@Composable
fun ErrorState(
    error: GatewayError,
    onCobaLagi: () -> Unit,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier.fillMaxWidth().padding(32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        Text("Gagal memuat", style = MaterialTheme.typography.titleLarge)
        Text(
            pesanGalat(error),
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center
        )
        Button(onClick = onCobaLagi) { Text("Coba lagi") }
    }
}

File: mobile/customer-ios-app/SukaShawarma/UI/Components/States.swift
import SwiftUI

/// Padanan `ui/components/States.kt`.
struct MemuatState: View {
    var body: some View {
        ProgressView()
            .tint(.sukaBrown)
            .frame(maxWidth: .infinity)
            .padding(32)
    }
}

struct EmptyState: View {
    let judul: String
    let penjelasan: String
    var tombol: (label: String, aksi: () -> Void)?

    var body: some View {
        VStack(spacing: 8) {
            Text(judul).font(SukaFont.titleLarge).foregroundStyle(Color.sukaInk)
            Text(penjelasan)
                .font(SukaFont.bodyMedium)
                .foregroundStyle(Color.sukaBody)
                .multilineTextAlignment(.center)
            if let tombol {
                TombolUtama(label: tombol.label, aksi: tombol.aksi).padding(.top, 8)
            }
        }
        .frame(maxWidth: .infinity)
        .padding(32)
    }
}

/// Menerjemahkan galat gateway jadi kalimat yang bisa ditindak pelanggan.
/// Satu-satunya tempat kalimat ini dikarang — layar tidak boleh membuat sendiri.
func pesanGalat(_ galat: GatewayError) -> String {
    switch galat {
    case .jaringan: "Tidak bisa terhubung. Periksa koneksi internetmu, lalu coba lagi."
    case .server: "Layanan sedang bermasalah. Coba lagi sebentar lagi."
    case .sesiTidakSah: "Sesimu sudah berakhir. Masuk lagi untuk melanjutkan."
    case .kode(_, let pesan): pesan
    }
}

struct ErrorState: View {
    let galat: GatewayError
    let onCobaLagi: () -> Void

    var body: some View {
        VStack(spacing: 12) {
            Text("Gagal memuat").font(SukaFont.titleLarge).foregroundStyle(Color.sukaInk)
            Text(pesanGalat(galat))
                .font(SukaFont.bodyMedium)
                .foregroundStyle(Color.sukaBody)
                .multilineTextAlignment(.center)
            TombolUtama(label: "Coba lagi", aksi: onCobaLagi)
        }
        .frame(maxWidth: .infinity)
        .padding(32)
    }
}

/// Tombol pil cokelat — padanan `Button` Material3 dengan tema Suka.
struct TombolUtama: View {
    let label: String
    let aksi: () -> Void

    var body: some View {
        Button(action: aksi) {
            Text(label)
                .font(SukaFont.jakarta(14, weight: .bold))
                .foregroundStyle(.white)
                .padding(.horizontal, 24).padding(.vertical, 12)
                .background(Color.sukaBrown, in: Capsule())
        }
        .buttonStyle(.mentul())
    }
}
